source/simulator/assembly.cc

```diff
@@ -2358,27 +637,79 @@ namespace aspect
-                                std_cxx11::_5));
+      {
+        assemblers->local_assemble_stokes_system
+        .connect (std_cxx11::bind(&aspect::StokesAssembler<dim>::local_assemble_stokes_incompressible,
+                                  std_cxx11::cref (*stokes_assembler),
+                                  // discard cell,
+                                  std_cxx11::_2,
+                                  std_cxx11::_3,
+                                  std_cxx11::_4,
+                                  std_cxx11::_5));
+
+        if (material_model->is_compressible())
+          assemblers->local_assemble_stokes_system
+          .connect (std_cxx11::bind(&aspect::StokesAssembler<dim>::local_assemble_stokes_compressible_diffusion,
+                                    std_cxx11::cref (*stokes_assembler),
+                                    // discard cell,
+                                    std_cxx11::_2,
+                                    std_cxx11::_3,
+                                    std_cxx11::_4,
+                                    std_cxx11::_5));
+
+        if (parameters.formulation_mass_conservation ==
+            Parameters<dim>::FormulationMassConservation::implicit_reference_density_profile)
+          assemblers->local_assemble_stokes_system
+          .connect (std_cxx11::bind(&aspect::StokesAssembler<dim>::local_assemble_stokes_mass_implicit_reference_density,
+                                    std_cxx11::cref (*stokes_assembler),
+                                    // discard cell,
+                                    std_cxx11::_2,
+                                    std_cxx11::_3,
+                                    std_cxx11::_4,
+                                    std_cxx11::_5,
+                                    std_cxx11::cref (this->parameters)));
+        else if (parameters.formulation_mass_conservation ==
+                 Parameters<dim>::FormulationMassConservation::reference_density_profile)
+          {
+            assemblers->local_assemble_stokes_system
+            .connect (std_cxx11::bind(&aspect::StokesAssembler<dim>::local_assemble_stokes_mass_reference_density,
+                                      std_cxx11::cref (*stokes_assembler),
+                                      // discard cell,
+                                      std_cxx11::_2,
+                                      std_cxx11::_3,
+                                      std_cxx11::_4,
+                                      std_cxx11::_5,
+                                      std_cxx11::cref (this->parameters)));
+          }
+        else if (parameters.formulation_mass_conservation ==
+                 Parameters<dim>::FormulationMassConservation::incompressible
+                 || (parameters.formulation_mass_conservation ==
+                     Parameters<dim>::FormulationMassConservation::ask_material_model
+                     && !material_model->is_compressible()))
```

Comment:
Is `ask_material_model` allowed here because the formulation does not matter if the material model says that it's not compressible anyway?

I would prefer if we simply asserted that we never encounter `ask_material_model` at all...